provider/aws_sd.go

```diff
@@ -0,0 +1,626 @@
+			newEndpoint.RecordType = endpoint.RecordTypeCNAME
+			newEndpoint.Targets = append(newEndpoint.Targets, aws.StringValue(inst.Attributes["AWS_INSTANCE_CNAME"]))
+
+			// ALIAS
+		} else if inst.Attributes["AWS_ALIAS_DNS_NAME"] != nil {
+			newEndpoint.RecordType = endpoint.RecordTypeCNAME
+			newEndpoint.Targets = append(newEndpoint.Targets, aws.StringValue(inst.Attributes["AWS_ALIAS_DNS_NAME"]))
+
+			// IP-based target
+		} else if inst.Attributes["AWS_INSTANCE_IPV4"] != nil {
+			newEndpoint.RecordType = endpoint.RecordTypeA
+			newEndpoint.Targets = append(newEndpoint.Targets, aws.StringValue(inst.Attributes["AWS_INSTANCE_IPV4"]))
+		} else {
+			return nil, fmt.Errorf("FATAL ERROR: Instance not valid \"%v\"", inst)
+		}
+	}
+
+	return newEndpoint, nil
+}
+
+// ApplyChanges applies Kubernetes changes in endpoints to AWS API
+func (p *AWSSDProvider) ApplyChanges(changes *plan.Changes) error {
+	// return early if there is nothing to change
+	if len(changes.Create) == 0 && len(changes.Delete) == 0 && len(changes.UpdateNew) == 0 {
+		log.Info("All records are already up to date")
+		return nil
+	}
+
+	// convert updates to delete and create operation if applicable (updates not supported)
+	creates, deletes := p.updatesToCreates(changes)
+	changes.Delete = append(changes.Delete, deletes...)
+	changes.Create = append(changes.Create, creates...)
+
+	namespaces, err := p.ListNamespaces()
+	if err != nil {
+		return err
+	}
+
+	err = p.submitDeletes(namespaces, changes.Delete)
+	if err != nil {
+		return err
+	}
+
+	err = p.submitCreates(namespaces, changes.Create)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (p *AWSSDProvider) updatesToCreates(changes *plan.Changes) (creates []*endpoint.Endpoint, deletes []*endpoint.Endpoint) {
+	for i, old := range changes.UpdateOld {
+		current := changes.UpdateNew[i]
+
+		if !reflect.DeepEqual(old.Targets, current.Targets) || old.DNSName != current.DNSName {
+			// when targets or DNS name differ the old instances need to be de-registered first
+			deletes = append(deletes, old)
+		}
+
+		// always register (or re-register) instance with the current data
+		creates = append(creates, current)
+	}
+
+	return creates, deletes
+}
+
+func (p *AWSSDProvider) submitCreates(namespaces []*sd.NamespaceSummary, changes []*endpoint.Endpoint) error {
+	changesByNamespaceID := p.changesByNamespaceID(namespaces, changes)
+
+	for nsID, changeList := range changesByNamespaceID {
+		services, err := p.ListServicesByNamespaceID(aws.String(nsID))
+		if err != nil {
+			return err
+		}
+
+		for _, ch := range changeList {
+			_, srvName := p.parseHostname(ch.DNSName)
+
+			srv := services[srvName]
+			summary := serviceToServiceSummary(srv)
+			if srv == nil {
+				// when service is missing create a new one
+				summary, err = p.CreateService(&nsID, &srvName, ch)
+				if err != nil {
+					return err
+				}
+				// update local list of services
+				services, err = p.ListServicesByNamespaceID(aws.String(nsID))
```

Comment:
doing this could easily lead to throttling, I am not entirely sure why we need to query Route53 for more information, when we do submitChanges, can we not do something similar to aws.go implementation ?

Reply:
Sorry, you were right. This call is not necessary. I will update it on Monday.